lib/Twig/Node/Expression/TemplateBlockReference.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) 2016 Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/**
+ * Represents a block call node on a template.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+class Twig_Node_Expression_TemplateBlockReference extends Twig_Node_Expression
+{
+    public function __construct(Twig_Node $template, Twig_Node $name, Twig_Node $context = null, $lineno)
+    {
+        $nodes = array('template' => $template, 'name' => $name);
+        if (null !== $context) {
+            $nodes['context'] = $context;
+        }
+
+        parent::__construct($nodes, array('output' => false), $lineno);
+    }
+
+    public function compile(Twig_Compiler $compiler)
+    {
+        if ($this->getAttribute('output')) {
+            $compiler
+                ->addDebugInfo($this)
+                ->write('$this->env->load(')
```

Comment:
Technically, the compiled code could avoid using the Twig_TemplateWrapper layer, as you don't need the output buffer cleanup logic here (it will be cleaned by the rendering of the template using the function already when the exception bubbles to it). This will avoid this level of call stack (and of try/catch). And the Twig code can relies on the Twig internal API.

Btw, this is why rendering blocks directly in the form theming was safe regarding the output buffer: it was always called inside a Twig call cleaning the buffer (unless someone was doing a crazy thing to call the TwigFormRenderer from a different templating engine than Twig). The email rendering use case was the unsafe one, but this one will not rely on ``template_block``